docs/backend-system/building-backends/08-migrating.md

```diff
@@ -7,3 +7,238 @@ description: How to migrate existing backends to the new backend system
+section.
+
+Sometimes though, type errors can be reported on the newly added line, saying
+that parts of the `PluginEnvironment` type do not match. This happens when the
+environment was changed from the defaults, perhaps with your own custom
+additions. If this is the case in your installation, you still aren't out of
+luck - you can build a customized `legacyPlugin` function.
+
+```diff
+ import { createBackend } from '@backstage/backend-defaults';
+-import { legacyPlugin } from '@backstage/backend-common';
++import { makeLegacyPlugin, loggerToWinstonLogger } from '@backstage/backend-common';
++import { coreServices } from '@backstage/backend-plugin-api';
+
++const legacyPlugin = makeLegacyPlugin(
++  {
++    cache: coreServices.cache,
++    config: coreServices.config,
++    database: coreServices.database,
++    discovery: coreServices.discovery,
++    logger: coreServices.logger,
++    permissions: coreServices.permissions,
++    scheduler: coreServices.scheduler,
++    tokenManager: coreServices.tokenManager,
++    reader: coreServices.urlReader,
++    identity: coreServices.identity,
++    // ... and your own additions
++  },
++  {
++    logger: log => loggerToWinstonLogger(log),
++  },
++);
+
+ const backend = createBackend();
+ backend.add(legacyPlugin('todo', import('./plugins/todo')));
+ backend.start();
+```
+
+The first argument to `makeLegacyPlugin` is the mapping from environment keys to
+references to actual [backend system services](../architecture/03-services.md).
+The second argument allows you to "tweak" the types of those services to
+something more fitting to your env. For example, you'll see that the logger
+service API type was changed from the raw Winston logger of old, to a different,
+custom API, so we use a helper function to transform that particular one.
+
+To make additions as mentioned above to the environment, you will start to get
+into the weeds of how the backend system wiring works. You'll need to have a
+service reference, and a service factory that performs the actual creation of
+your service. For now, let's add them directly in the same file, and then you
+can move them out into a separate file at a later time.
+
+In this example, we'll assume that your added environment item is named "example".
+
+```diff
+-import { coreServices } from '@backstage/backend-plugin-api';
++import { coreServices, createServiceFactory } from '@backstage/backend-plugin-api';
++import { ExampleApi, ExampleImpl } from '<somewhere>';
+
++const exampleServiceRef = createServiceRef<ExampleApi>({
++  id: 'example',
++  scope: 'plugin', // can be 'root' or 'plugin'
++  defaultFactory: async service => createServiceFactory({
++    service,
++    // optional; just an example of how dependencies work
++    deps: {
++      logger: coreServices.logger,
++    },
++    // here instances of the declared dependencies are injected
++    async factory({ logger }) {
++      // return your ExampleImpl
++    },
++  }),
++});
```

Comment:
Perhaps it would be better to completely defer this example to the service architecture docs? Just mentioning and linking to the appropriate section here. It will likely be very rare that people need to do this and it does take up a lot of space. Code examples grabs people's attention when reading docs too, so some might be led astray here x)

Reply:
aight moving it there